File: graphics/examplesPython/probe.py
#!/usr/local/bin/python

from libVTKCommonPython import *
from libVTKGraphicsPython import *

#catch  load vtktcl 
# get the interactor ui
#source ../../examplesTcl/vtkInt.tcl
#source ../../examplesTcl/colors.tcl
from colors import *
# Create the RenderWindow, Renderer and both Actors
#
ren = vtkRenderer()
renWin = vtkRenderWindow()
renWin.AddRenderer(ren)
iren = vtkRenderWindowInteractor()
iren.SetRenderWindow(renWin)

# cut data
pl3d = vtkPLOT3DReader()
pl3d.SetXYZFileName("../../../vtkdata/combxyz.bin")
pl3d.SetQFileName("../../../vtkdata/combq.bin")
pl3d.SetScalarFunctionNumber(100)
pl3d.SetVectorFunctionNumber(202)
pl3d.Update()
plane = vtkPlane()
plane.SetOrigin(pl3d.GetOutput().GetCenter())
plane.SetNormal(-0.287,0,0.9579)
planeCut = vtkCutter()
planeCut.SetInput(pl3d.GetOutput())
planeCut.SetCutFunction(plane)
probe = vtkProbeFilter()
probe.SetInput(planeCut.GetOutput())
probe.SetSource(pl3d.GetOutput())
cutMapper = vtkDataSetMapper()
cutMapper.SetInput(probe.GetOutput())
cutMapper.SetScalarRange(  \
	pl3d.GetOutput().GetPointData().GetScalars().GetRange() )
cutActor = vtkActor()
cutActor.SetMapper(cutMapper)

#extract plane
compPlane = vtkStructuredGridGeometryFilter()
compPlane.SetInput(pl3d.GetOutput())
compPlane.SetExtent(0,100,0,100,9,9)
planeMapper = vtkPolyDataMapper()
planeMapper.SetInput(compPlane.GetOutput())
planeMapper.ScalarVisibilityOff()
planeActor = vtkActor()
planeActor.SetMapper(planeMapper)
planeActor.GetProperty().SetRepresentationToWireframe()
planeActor.GetProperty().SetColor(0,0,0)

#outline
outline = vtkStructuredGridOutlineFilter()
outline.SetInput(pl3d.GetOutput())
outlineMapper = vtkPolyDataMapper()
outlineMapper.SetInput(outline.GetOutput())
outlineActor = vtkActor()
outlineActor.SetMapper(outlineMapper)
outlineProp=outlineActor.GetProperty()
outlineProp.SetColor(0,0,0)

# Add the actors to the renderer, set the background and size
#
ren.AddActor(outlineActor)
ren.AddActor(planeActor)
ren.AddActor(cutActor)
ren.SetBackground(1,1,1)
renWin.SetSize(500,500)

cam1=ren.GetActiveCamera()
cam1.SetClippingRange(3.95297,50)
cam1.SetFocalPoint(9.71821,0.458166,29.3999)
cam1.SetPosition(2.7439,-37.3196,38.7167)
cam1.ComputeViewPlaneNormal()
cam1.SetViewUp(-0.16123,0.264271,0.950876)
iren.Initialize()

# render the image
#

#renWin SetFileName "probe.tcl.ppm"
#renWin SaveImageAsPPM

# prevent the tk window from showing up then start the event loop
#wm withdraw .



iren.Start()

File: graphics/examplesCxx/probe.cxx
#include "vtkRenderer.h"
#include "vtkRenderWindow.h"
#include "vtkRenderWindowInteractor.h"
#include "vtkPLOT3DReader.h"
#include "vtkStructuredGridOutlineFilter.h"
#include "vtkPolyDataMapper.h"
#include "vtkActor.h"
#include "vtkCursor3D.h"
#include "vtkProbeFilter.h"
#include "vtkConeSource.h"
#include "vtkGlyph3D.h"

#include "SaveImage.h"

int main( int argc, char *argv[] )
{
  vtkRenderer *renderer = vtkRenderer::New();
  vtkRenderWindow *renWin = vtkRenderWindow::New();
    renWin->AddRenderer(renderer);
  vtkRenderWindowInteractor *iren = vtkRenderWindowInteractor::New();
    iren->SetRenderWindow(renWin);

  // read data
  vtkPLOT3DReader *reader = vtkPLOT3DReader::New();
    //reader->DebugOn();
    reader->SetXYZFileName("../../../vtkdata/combxyz.bin");
    reader->SetQFileName("../../../vtkdata/combq.bin");
    reader->SetFileFormat(VTK_WHOLE_SINGLE_GRID_NO_IBLANKING);
    reader->Update(); //force a read to occur

  // create outline
  vtkStructuredGridOutlineFilter *outlineF = vtkStructuredGridOutlineFilter::New();
    outlineF->SetInput(reader->GetOutput());
  vtkPolyDataMapper *outlineMapper = vtkPolyDataMapper::New();
    outlineMapper->SetInput(outlineF->GetOutput());
  vtkActor *outline = vtkActor::New();
    outline->SetMapper(outlineMapper);
    outline->GetProperty()->SetColor(0,0,0);

  // create cursor
  vtkCursor3D *cursor = vtkCursor3D::New();
    cursor->SetFocalPoint(reader->GetOutput()->GetCenter());
    cursor->SetModelBounds(reader->GetOutput()->GetBounds());
    cursor->AxesOn();
    cursor->OutlineOff();
    cursor->XShadowsOff();
    cursor->YShadowsOff();
    cursor->ZShadowsOff();
  vtkPolyDataMapper *cursorMapper = vtkPolyDataMapper::New();
    cursorMapper->SetInput(cursor->GetOutput());
  vtkActor *cursorActor = vtkActor::New();
    cursorActor->SetMapper(cursorMapper);
    cursorActor->GetProperty()->SetColor(1,0,0);

  // create probe
  vtkProbeFilter *probe = vtkProbeFilter::New();
    probe->SetSource(reader->GetOutput());
    probe->SetInput(cursor->GetFocus());

  // create a cone geometry for glyph
  vtkConeSource *cone = vtkConeSource::New();
    cone->SetResolution(16);
    cone->SetRadius(0.25);

  // create glyph
  vtkGlyph3D *glyph = vtkGlyph3D::New();
    glyph->SetInput(probe->GetOutput());
    glyph->SetSource(cone->GetOutput());
    glyph->SetVectorModeToUseVector();
    glyph->SetScaleModeToScaleByScalar();
    glyph->SetScaleFactor(10);
  vtkPolyDataMapper *glyphMapper = vtkPolyDataMapper::New();
    glyphMapper->SetInput(glyph->GetOutput());
  vtkActor *glyphActor = vtkActor::New();
    glyphActor->SetMapper(glyphMapper);

  renderer->AddActor(outline);
  renderer->AddActor(cursorActor);
  renderer->AddActor(glyphActor);
  renderer->SetBackground(1.0,1.0,1.0);
  renWin->SetSize(300,300);
  renWin->Render();

  SAVEIMAGE( renWin );

  iren->Start();

  // Clean up
  renderer->Delete();
  renWin->Delete();
  iren->Delete();
  reader->Delete();
  outlineF->Delete();
  outlineMapper->Delete();
  outline->Delete();
  cursor->Delete();
  cursorMapper->Delete();
  cursorActor->Delete();
  probe->Delete();
  cone->Delete();
  glyph->Delete();
  glyphMapper->Delete();
  glyphActor->Delete();
}
